libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp

```diff
@@ -287,9 +317,87 @@ namespace meshkernel
+
+        /// @brief Compute the node indices in row-major order (n runs fastest).
+        /// @returns The  mapping (m and n indices for each node)
+        [[nodiscard]] std::vector<Point> ComputeFaceCenters() const;
+
+        /// @brief Computes the valid grid faces
+        void ComputeGridFacesMask();
+
+        /// @brief Compute the node indices in row-major order (n runs fastest).
+        /// @returns The  mapping (m and n indices for each node)
+        [[nodiscard]] std::vector<CurvilinearGridNodeIndices> ComputeNodeIndices() const;
+
+        /// @brief Compute the edge indices.
+        /// @returns The  mapping (m and n indices for each node of the edge)
+        [[nodiscard]] std::vector<CurvilinearEdgeNodeIndices> ComputeEdgeIndices() const;
+
+        /// @brief Compute the face indices.
+        /// @returns The  mapping (m and n indices for each node of the face)
+        [[nodiscard]] std::vector<CurvilinearFaceNodeIndices> ComputeFaceIndices() const;
+
+        /// @brief Remove invalid nodes.
+        /// This function is recursive
+        /// @param[in] invalidNodesToRemove Whether there are still invalid nodes to remove
+        void RemoveInvalidNodes(bool invalidNodesToRemove);
+
+        /// @brief Set the m_nodesRTreeRequiresUpdate flag
+        /// @param[in] value The value of the flag
+        void SetNodesRTreeRequiresUpdate(bool value) { m_nodesRTreeRequiresUpdate = value; }
+
+        /// @brief Set the m_edgesRTreeRequiresUpdate flag
+        /// @param[in] value The value of the flag
+        void SetEdgesRTreeRequiresUpdate(bool value) { m_edgesRTreeRequiresUpdate = value; }
+
+        /// @brief Set the m_facesRTreeRequiresUpdate flag
+        /// @param[in] value The value of the flag
+        void SetFacesRTreeRequiresUpdate(bool value) { m_facesRTreeRequiresUpdate = value; }
+
+        Projection m_projection;                               ///< The curvilinear grid projection
         lin_alg::Matrix<Point> m_gridNodes;                    ///< Member variable storing the grid
         lin_alg::Matrix<bool> m_gridFacesMask;                 ///< The mask of the grid faces (true/false)
         lin_alg::Matrix<NodeType> m_gridNodesTypes;            ///< The grid node types
         std::vector<CurvilinearGridNodeIndices> m_gridIndices; ///< The original mapping of the flatten nodes in the curvilinear grid
+
+        // RTrees
+        bool m_nodesRTreeRequiresUpdate = true;                            ///< m_nodesRTree requires an update
+        bool m_edgesRTreeRequiresUpdate = true;                            ///< m_edgesRTree requires an update
+        bool m_facesRTreeRequiresUpdate = true;                            ///< m_facesRTree requires an update
+        std::unordered_map<Location, std::unique_ptr<RTreeBase>> m_RTrees; ///< The RTrees to use
+        BoundingBox m_boundingBoxCache;                                    ///< Caches the last bounding box used for selecting the locations
```

Comment:
Could it be an idea to remove the trees from the curvilinear mesh completely?
Have functions that can create trees that have been pass as parameters.
The the algorithms that need the functionality provided by the tree can defined thier own trees and use them

Reply:
there is a separate task for this